pyro/contrib/gp/kernels/kernel.py

```diff
@@ -52,3 +68,238 @@ def _slice_input(self, X):
             return X.unsqueeze(1)
         else:
             raise ValueError("Input X must be either 1 or 2 dimensional.")
+
+    def __add__(self, other):
+        return self.add(other)
+
+    def __mul__(self, other):
+        return self.mul(other)
+
+    def __radd__(self, other):
+        return self.add(other)
+
+    def __rmul__(self, other):
+        return self.mul(other)
+
+    def add(self, other, name=None):
+        """
+        Returns a new kernel which acts like a sum/direct sum of ``self`` and ``other``.
+        """
+        return Sum(self, other, None)
+
+    def mul(self, other, name=None):
+        """
+        Returns a new kernel which acts like a product/tensor product of ``self`` and ``other``.
+        """
+        return Product(self, other, name)
+
+    def exp(self, name=None):
+        """
+        Creates a new kernel, which is an instance of Exponent kernel, according to:
+        ``k_new(x, z) = exp(k(x, z))``.
+        """
+        return Exponent(self, name=name)
+
+    def vertical_scale(self, vscaling_fn, name=None):
+        """
+        Creates a new kernel, which is an instance of VerticalScaling kernel, according to:
+        ``k_new(x, z) = f(x)k(x, z)f(z)``,
+        where ``f := vscaling_fn`` is a vertical scaling function.
+        """
+        return VerticalScaling(self, vscaling_fn, name=name)
+
+    def warp(self, iwarping_fn=None, owarping_coef=None, name=None):
+        """
+        Creates a new kernel, which is an instance of Warping kernel, according to:
+        ``k_new(x, z) = q(k(f(x), f(z)))``,
+        where ``f := iwarping_fn`` is a input warping function, and ``q`` is a polynomial
+        with non-negative coefficients ``owarping_coef``.
+        """
+        return Warping(self, iwarping_fn, owarping_coef, name)
+
+    def get_subkernel(self, name):
+        """
+        Returns the subkernel corresponding to ``name``.
+        """
+        if name in self._subkernels:
+            return self._subkernels[name]
+        else:
+            raise KeyError("There is no subkernel with the name '{}'.".format(name))
+
+
+class Combination(Kernel):
+    """
+    Base class for Sum and Product kernels.
+
+    :param pyro.contrib.gp.kernels.Kernel kernel0: First kernel to combine.
+    :param pyro.contrib.gp.kernels.Kernel or numbers.Number kernel1: Second kernel to combine.
+    """
+
+    def __init__(self, kern0, kern1, name=None):
+        if not (isinstance(kern0, Kernel) and
```

Comment:
nit: This would be clearer as a pair of checks:
```py
if not isinstance(kern0, Kernel):
    raise TypeError(...)
if not isinstance(kern1, (Kernel, numbers.Number)):
    raise TypeError(...)
```